server/models/Board.js

```diff
@@ -1,29 +1,108 @@
 const mongoose = require("mongoose");
+const Column = require("../models/Column");
+const Card = require("../models/Card");
 
-const boardSchema = new mongoose.Schema({
-  name: {
-    type: String,
-    required: true,
+const boardSchema = new mongoose.Schema(
+  {
+    name: {
+      type: String,
+      required: true,
+    },
+    columns: [
+      { type: mongoose.Schema.Types.ObjectId, ref: "column", required: false },
+    ],
+    user: {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: "user",
+      required: true,
+    },
   },
-  columns: {
-    type: Array,
-    required: false,
-  },
-  user: {
-    type: mongoose.Types.ObjectId,
-    ref: "user",
-    required: true,
-  },
-  createdAt: {
-    type: Date,
-    default: Date.now,
-  },
-});
+  { timestamps: true }
+);
 
+// Method that will remove the user's password before returning the board; this can be redone to ensure we don't pop user
 boardSchema.methods.removePassword = function () {
-  const object = this.toObject();
-  delete object.user.password;
-  return object;
+  const board = this.toObject();
+  delete board.user.password;
+  return board;
+};
+
+boardSchema.methods.addColumn = async function (
+  /** @type {string} */ side,
+  /** @type {string} */ columnName
+) {
+  const newColumn = await Column.create({
+    name: columnName,
+    boardId: this._id,
+  });
+
+  side === "left"
+    ? this.columns.unshift(newColumn)
+    : this.columns.push(newColumn);
+
+  await this.populate();
+  await this.save();
+};
+
+boardSchema.methods.deleteColumn = async function (
+  /** @type {string} */ columnId
+) {
+  const colIndex = this.columns.findIndex((column) => column._id === columnId);
+  this.columns.splice(colIndex, 1);
+  await this.populate();
+  await this.save();
+};
+
+boardSchema.methods.updateName = async function (
+  /** @type {string} */ newName
+) {
+  this.name = newName;
+
+  await this.populate();
+  await this.save();
+};
+
+boardSchema.methods.createTemplateBoard = async function () {
+  const notStarted = await Column.create({
+    name: "Not Started",
+    boardId: this._id,
+  });
+  const inProgress = await Column.create({
+    name: "In Progress",
+    boardId: this._id,
+  });
+  const completed = await Column.create({
+    name: "Completed",
+    boardId: this._id,
+  });
+
+  const card1 = await Card.create({
+    title: "Not Started Card 1",
+    tag: "green",
+    columnId: notStarted._id,
+  });
+  const card2 = await Card.create({
+    title: "In Progress Card 1",
+    tag: "blue",
+    columnId: inProgress._id,
+  });
+  const card3 = await Card.create({
+    title: "Completed Card 1",
+    tag: "red",
+    columnId: notStarted._id,
+  });
```

Comment:
I don't think giving a user 3 cards really is necessary, just one in a column will be fine for a user to start off.

Reply:
Fair enough; it was useful for testing and such, but I agree this would not be desirable in actual use. I updated it to 3 empty columns (not started/in progress/complete) for now, but we can change it before the presentation :)